File: scripts/build_address_index.py
#!/usr/bin/env python3
"""
Build Address Search Index for "Analyse Your Home" feature.

Creates a single `address_search_index` collection in Gold_Coast DB with
lightweight address records from all suburb collections. This enables
fast autocomplete search (~100-200ms) instead of scanning 80+ collections.

Usage:
    python3 scripts/build_address_index.py
    python3 scripts/build_address_index.py --drop  # rebuild from scratch
"""
import os
import sys
import time
import argparse
from datetime import datetime
from pymongo import MongoClient
from pymongo.errors import OperationFailure, BulkWriteError

INDEX_COLLECTION = 'address_search_index'

TARGET_MARKET_SUBURBS = [
    'robina', 'mudgeeraba', 'varsity_lakes', 'carrara',
    'reedy_creek', 'burleigh_waters', 'merrimac', 'worongary',
]

SKIP_COLLECTIONS = {
    INDEX_COLLECTION, 'system.profile', 'suburb_median_prices',
    'suburb_statistics', 'change_detection_snapshots',
    'precomputed_indexed_prices', 'precomputed_motion_data',
    'precomputed_property_type_race',
}


def suburb_display_name(coll_name):
    return ' '.join(w.capitalize() for w in coll_name.split('_'))


def format_street_address(doc):
    parts = []
    if doc.get('STREET_NO_1'):
        parts.append(str(doc['STREET_NO_1']))
    if doc.get('STREET_NAME'):
        parts.append(doc['STREET_NAME'].title())
    if doc.get('STREET_TYPE'):
        parts.append(doc['STREET_TYPE'].title())
    return ' '.join(parts)


def cosmos_retry(fn, *args, max_retries=5, **kwargs):
    for attempt in range(max_retries):
        try:
            return fn(*args, **kwargs)
        except BulkWriteError:
            # Duplicate key errors are expected in resume mode — ignore
            return
        except OperationFailure as e:
            if '16500' in str(e) or '429' in str(e):
                wait = min(2 ** attempt * 2, 30)
                print(f"  429 rate limit, waiting {wait}s (attempt {attempt+1})...", flush=True)
                time.sleep(wait)
            else:
                raise
    raise Exception(f"Failed after {max_retries} retries")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--drop', action='store_true', help='Drop and rebuild index')
    args = parser.parse_args()

    uri = os.environ.get('COSMOS_CONNECTION_STRING')
    if not uri:
        print("COSMOS_CONNECTION_STRING not set")
        sys.exit(1)

    client = MongoClient(uri, retryWrites=False, tls=True, tlsAllowInvalidCertificates=True)
    db = client['Gold_Coast']
    print(f"Build Address Index -- {datetime.now().strftime('%Y-%m-%d %H:%M')}", flush=True)

    idx_coll = db[INDEX_COLLECTION]

    if args.drop:
        print(f"Dropping {INDEX_COLLECTION}...", flush=True)
        db.drop_collection(INDEX_COLLECTION)
        time.sleep(2)

    # Check which suburbs already indexed
    already_indexed = set()
    try:
        pipeline = [{'$group': {'_id': '$suburb_key'}}]
        for doc in idx_coll.aggregate(pipeline):
            already_indexed.add(doc['_id'])
        if already_indexed:
            print(f"Already indexed: {len(already_indexed)} suburbs", flush=True)
    except Exception:
        pass

    # Get all suburb collections
    all_colls = db.list_collection_names()
    suburb_colls = [c for c in all_colls if c not in SKIP_COLLECTIONS and not c.startswith('system.')]

    # Target market first
    target = [c for c in suburb_colls if c in TARGET_MARKET_SUBURBS]
    others = sorted([c for c in suburb_colls if c not in TARGET_MARKET_SUBURBS])
    ordered = target + others

    print(f"Found {len(ordered)} suburb collections", flush=True)

    total_inserted = 0

    for suburb_name in ordered:
        if suburb_name in already_indexed:
            print(f"  Skip {suburb_display_name(suburb_name)} (already indexed)", flush=True)
            continue

        coll = db[suburb_name]
        is_target = suburb_name in TARGET_MARKET_SUBURBS

        projection = {
            '_id': 1, 'complete_address': 1,
            'STREET_NO_1': 1, 'STREET_NAME': 1, 'STREET_TYPE': 1,
            'LOCALITY': 1, 'POSTCODE': 1, 'PROPERTY_TYPE': 1,
            'images': 1, 'lot_size_sqm': 1,
            'scraped_data.bedrooms': 1, 'scraped_data.bathrooms': 1,
            'scraped_data.car_spaces': 1, 'scraped_data.features': 1,
            'bedrooms': 1, 'bathrooms': 1,
        }

        cursor = coll.find({}, projection).batch_size(100)
        batch = []
        count = 0
        BATCH_SIZE = 50

        for doc in cursor:
            addr = doc.get('complete_address')
            if not addr:
                continue

            scraped = doc.get('scraped_data', {})
            images = doc.get('images', [])
            if not isinstance(images, list):
                images = []

            bedrooms = doc.get('bedrooms') or scraped.get('bedrooms')
            bathrooms = doc.get('bathrooms') or scraped.get('bathrooms')
            car_spaces = scraped.get('car_spaces')

            index_doc = {
                'address': addr.upper(),
                'address_display': format_street_address(doc),
                'street_no': str(doc.get('STREET_NO_1', '')),
                'street_name': (doc.get('STREET_NAME') or '').upper(),
                'street_type': (doc.get('STREET_TYPE') or '').upper(),
                'source_id': doc['_id'],
                'suburb_key': suburb_name,
                'suburb': suburb_display_name(suburb_name),
                'postcode': doc.get('POSTCODE', ''),
                'property_type': doc.get('PROPERTY_TYPE', 'Residential'),
                'lot_size_sqm': doc.get('lot_size_sqm'),
                'has_images': len(images) > 0,
                'image_count': len(images),
                'bedrooms': bedrooms,
                'bathrooms': bathrooms,
                'car_spaces': car_spaces,
                'is_target_market': is_target,
            }

            batch.append(index_doc)
            count += 1

            if len(batch) >= BATCH_SIZE:
                cosmos_retry(idx_coll.insert_many, batch, ordered=False)
                total_inserted += len(batch)
                batch = []
                time.sleep(0.3)

        if batch:
            cosmos_retry(idx_coll.insert_many, batch, ordered=False)
            total_inserted += len(batch)

        print(f"  {suburb_display_name(suburb_name)}: {count} indexed", flush=True)
        time.sleep(1)

    # Create indexes
    print("\nCreating indexes...", flush=True)
    try:
        cosmos_retry(idx_coll.create_index,
            [('street_name', 1), ('street_no', 1), ('suburb_key', 1)],
            name='street_search_idx')
        print("  Created street_search_idx", flush=True)
    except Exception as e:
        print(f"  street_search_idx: {e}", flush=True)

    try:
        cosmos_retry(idx_coll.create_index, [('suburb_key', 1)], name='suburb_idx')
        print("  Created suburb_idx", flush=True)
    except Exception as e:
        print(f"  suburb_idx: {e}", flush=True)

    try:
        cosmos_retry(idx_coll.create_index, [('address', 1)], name='address_idx')
        print("  Created address_idx", flush=True)
    except Exception as e:
        print(f"  address_idx: {e}", flush=True)

    try:
        cosmos_retry(idx_coll.create_index,
            [('is_target_market', -1), ('suburb_key', 1)],
            name='target_market_idx')
        print("  Created target_market_idx", flush=True)
    except Exception as e:
        print(f"  target_market_idx: {e}", flush=True)

    final_count = idx_coll.estimated_document_count()
    print(f"\nDone! {total_inserted} new records inserted. Total: {final_count}", flush=True)
    client.close()


if __name__ == '__main__':
    main()

File: scripts/build-address-index.mjs
/**
 * Build Address Search Index — One-time migration script
 * 
 * Last Updated: 10/02/2026, 1:27 PM (Monday) — Brisbane Time
 * 
 * Edit History:
 * - 10/02/2026 1:27 PM: Changed to RESUME mode (no longer drops collection)
 *   - Skips suburbs already in the index
 *   - Processes target market suburbs FIRST for faster time-to-value
 *   - Increased delay between batches to 300ms to avoid 429 rate limits
 * - 10/02/2026 12:37 PM: Initial creation
 *   - Reads ALL suburb collections from Gold_Coast database
 *   - Extracts lightweight address records (address, suburb, source_id, postcode, etc.)
 *   - Writes them into a single `address_search_index` collection in Gold_Coast database
 *   - Creates a text index on the `address` field for fast $text search
 *   - Also creates a compound regex-friendly index on address components
 *   - ~200k properties → single indexed collection = 100-200ms search instead of 3-26s
 * 
 * Description:
 * This script solves the core performance problem with address search. Currently,
 * every search query scans 52 separate suburb collections using $regex (no index).
 * This script consolidates all addresses into ONE collection with proper indexes,
 * reducing search time from 3-26 seconds to ~100-200ms.
 * 
 * Usage:
 *   cd /Users/projects/Documents/Feilds_Website/01_Website && node scripts/build-address-index.mjs
 * 
 * Prerequisites:
 *   - COSMOS_CONNECTION_STRING environment variable set (or .env file in 01_Website)
 *   - Access to Gold_Coast database on Azure Cosmos DB
 * 
 * Safety:
 *   - Drops and recreates the `address_search_index` collection each run
 *   - Does NOT modify any source suburb collections
 *   - Safe to re-run at any time (idempotent)
 */

import { MongoClient } from 'mongodb';
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

// Load .env from 01_Website directory (no dotenv dependency needed)
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const envPath = join(__dirname, '..', '.env');
try {
  const envContent = readFileSync(envPath, 'utf-8');
  for (const line of envContent.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const eqIdx = trimmed.indexOf('=');
    if (eqIdx === -1) continue;
    const key = trimmed.slice(0, eqIdx).trim();
    let value = trimmed.slice(eqIdx + 1).trim();
    // Remove surrounding quotes if present
    if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
      value = value.slice(1, -1);
    }
    if (!process.env[key]) {
      process.env[key] = value;
    }
  }
  console.log('📄 Loaded .env from', envPath);
} catch (e) {
  console.log('⚠️  No .env file found at', envPath, '— using environment variables');
}

const INDEX_COLLECTION = 'address_search_index';

const TARGET_MARKET_SUBURBS = [
  'robina',
  'mudgeeraba',
  'varsity_lakes',
  'carrara',
  'reedy_creek',
  'burleigh_waters',
  'merrimac',
  'worongary',
];

/**
 * Format address for display
 * "5 FULHAM PLACE ROBINA QLD 4226" → "5 Fulham Place"
 */
function formatStreetAddress(doc) {
  const parts = [];
  if (doc.STREET_NO_1) parts.push(doc.STREET_NO_1);
  if (doc.STREET_NAME) {
    const name = doc.STREET_NAME.charAt(0).toUpperCase() + doc.STREET_NAME.slice(1).toLowerCase();
    parts.push(name);
  }
  if (doc.STREET_TYPE) {
    const type = doc.STREET_TYPE.charAt(0).toUpperCase() + doc.STREET_TYPE.slice(1).toLowerCase();
    parts.push(type);
  }
  return parts.join(' ');
}

/**
 * Convert suburb collection name to display name
 * "varsity_lakes" → "Varsity Lakes"
 */
function suburbDisplayName(collName) {
  return collName
    .split('_')
    .map(word => word.charAt(0).toUpperCase() + word.slice(1))
    .join(' ');
}

/**
 * Insert documents with retry logic for Cosmos DB 429 rate limits.
 * Retries up to 5 times with exponential backoff.
 */
async function insertWithRetry(collection, docs, maxRetries = 5) {
  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      await collection.insertMany([...docs], { ordered: false });
      // Delay between successful batches to stay under Cosmos DB rate limit
      await new Promise(r => setTimeout(r, 300));
      return;
    } catch (err) {
      const isRateLimit = err.code === 16500 || (err.message && err.message.includes('TooManyRequests'));
      if (isRateLimit && attempt < maxRetries - 1) {
        // Extract RetryAfterMs from error or use exponential backoff
        const retryMatch = err.message && err.message.match(/RetryAfterMs=(\d+)/);
        const retryMs = retryMatch ? parseInt(retryMatch[1]) : 100;
        const backoff = Math.max(retryMs, 200 * Math.pow(2, attempt));
        process.stdout.write(`  ⏳ Rate limited, waiting ${backoff}ms (attempt ${attempt + 1}/${maxRetries})...\r`);
        await new Promise(r => setTimeout(r, backoff));
        
        // For partial failures (some docs inserted), we need to figure out which ones failed
        // With ordered:false, insertMany inserts as many as possible
        // The error result tells us how many were inserted
        if (err.result && err.result.insertedCount > 0) {
          // Some were inserted, skip those
          return; // Accept partial success
        }
      } else {
        throw err;
      }
    }
  }
}

async function main() {
  const uri = process.env.COSMOS_CONNECTION_STRING;
  if (!uri) {
    console.error('❌ COSMOS_CONNECTION_STRING not set. Check your .env file.');
    process.exit(1);
  }

  console.log('🔗 Connecting to Azure Cosmos DB...');
  const client = new MongoClient(uri, {
    retryWrites: false,
    maxIdleTimeMS: 120000,
    serverSelectionTimeoutMS: 30000,
    socketTimeoutMS: 120000,
    connectTimeoutMS: 30000,
  });

  try {
    await client.connect();
    const db = client.db('Gold_Coast');
    console.log('✅ Connected to Gold_Coast database');

    // Get all suburb collections
    const collections = await db.listCollections().toArray();
    const suburbCollections = collections
      .map(c => c.name)
      .filter(name => !name.startsWith('system.') && name !== INDEX_COLLECTION);

    console.log(`📋 Found ${suburbCollections.length} suburb collections`);

    const indexCollection = db.collection(INDEX_COLLECTION);

    // RESUME MODE: Check which suburbs are already indexed
    let alreadyIndexed = new Set();
    try {
      const pipeline = [{ $group: { _id: '$suburb_key' } }];
      const existing = await indexCollection.aggregate(pipeline).toArray();
      alreadyIndexed = new Set(existing.map(e => e._id));
      console.log(`📊 Already indexed: ${alreadyIndexed.size} suburbs: ${[...alreadyIndexed].join(', ')}`);
    } catch (e) {
      console.log(`📝 ${INDEX_COLLECTION} collection is empty or doesn't exist yet`);
    }

    // Sort: target market suburbs FIRST, then alphabetical
    const targetFirst = suburbCollections.filter(c => TARGET_MARKET_SUBURBS.includes(c));
    const others = suburbCollections.filter(c => !TARGET_MARKET_SUBURBS.includes(c)).sort();
    const orderedCollections = [...targetFirst, ...others];

    let totalProcessed = 0;
    let totalInserted = 0;
    let skipped = 0;

    // Process each suburb collection
    for (const suburbName of orderedCollections) {
      // Skip already-indexed suburbs
      if (alreadyIndexed.has(suburbName)) {
        skipped++;
        console.log(`⏭️  Skipping ${suburbDisplayName(suburbName)} (already indexed)`);
        continue;
      }

      const coll = db.collection(suburbName);
      
      // Count documents
      const count = await coll.countDocuments();
      console.log(`\n📍 Processing ${suburbDisplayName(suburbName)} (${count} properties)...`);

      // Read all documents with just the fields we need
      const cursor = coll.find({}).project({
        _id: 1,
        complete_address: 1,
        STREET_NO_1: 1,
        STREET_NAME: 1,
        STREET_TYPE: 1,
        LOCALITY: 1,
        POSTCODE: 1,
        PROPERTY_TYPE: 1,
        images: 1,
        'scraped_data.bedrooms': 1,
        'scraped_data.bathrooms': 1,
        'scraped_data.car_spaces': 1,
      });

      const batch = [];
      const BATCH_SIZE = 50; // Small batches to avoid Cosmos DB 429 rate limits

      for await (const doc of cursor) {
        totalProcessed++;

        if (!doc.complete_address) continue;

        const isTargetMarket = TARGET_MARKET_SUBURBS.includes(suburbName);
        const images = Array.isArray(doc.images) ? doc.images : [];
        const scraped = doc.scraped_data || {};

        // Build the index document — lightweight, just what search needs
        const indexDoc = {
          // Search fields
          address: doc.complete_address.toUpperCase(),
          street_no: doc.STREET_NO_1 || '',
          street_name: (doc.STREET_NAME || '').toUpperCase(),
          street_type: (doc.STREET_TYPE || '').toUpperCase(),
          
          // Reference fields
          source_id: doc._id,
          suburb_key: suburbName,
          suburb: suburbDisplayName(suburbName),
          postcode: doc.POSTCODE || '',
          
          // Display fields (for autocomplete dropdown)
          property_type: doc.PROPERTY_TYPE || 'Residential',
          has_images: images.length > 0,
          image_count: images.length,
          bedrooms: scraped.bedrooms || null,
          bathrooms: scraped.bathrooms || null,
          car_spaces: scraped.car_spaces || null,
          is_target_market: isTargetMarket,
        };

        batch.push(indexDoc);

        // Insert in small batches with retry logic for Cosmos DB rate limits
        if (batch.length >= BATCH_SIZE) {
          await insertWithRetry(indexCollection, batch);
          totalInserted += batch.length;
          batch.length = 0;
          process.stdout.write(`  Inserted ${totalInserted} records...\r`);
        }
      }

      // Insert remaining
      if (batch.length > 0) {
        await insertWithRetry(indexCollection, batch);
        totalInserted += batch.length;
      }

      console.log(`  ✅ ${suburbDisplayName(suburbName)}: ${count} properties indexed`);
    }

    console.log(`\n📊 Total: ${totalProcessed} processed, ${totalInserted} indexed`);

    // Create indexes for fast searching
    console.log('\n🔧 Creating indexes...');

    // Index 1: Compound index on street_no + street_name for structured search
    // This is the fastest path: user types "5 Fulham" → exact match on street_no + prefix on street_name
    try {
      await indexCollection.createIndex(
        { street_name: 1, street_no: 1, suburb_key: 1 },
        { name: 'street_search_idx' }
      );
      console.log('  ✅ Created street_search_idx (street_name + street_no + suburb_key)');
    } catch (e) {
      console.log(`  ⚠️  street_search_idx: ${e.message}`);
    }

    // Index 2: Index on suburb_key for suburb-filtered searches
    try {
      await indexCollection.createIndex(
        { suburb_key: 1 },
        { name: 'suburb_idx' }
      );
      console.log('  ✅ Created suburb_idx');
    } catch (e) {
      console.log(`  ⚠️  suburb_idx: ${e.message}`);
    }

    // Index 3: Index on address field for regex prefix search
    // Cosmos DB doesn't support $text indexes, but a regular index on address
    // helps with prefix regex like /^5 FULHAM/i
    try {
      await indexCollection.createIndex(
        { address: 1 },
        { name: 'address_idx' }
      );
      console.log('  ✅ Created address_idx');
    } catch (e) {
      console.log(`  ⚠️  address_idx: ${e.message}`);
    }

    // Index 4: Target market priority index
    try {
      await indexCollection.createIndex(
        { is_target_market: -1, suburb_key: 1 },
        { name: 'target_market_idx' }
      );
      console.log('  ✅ Created target_market_idx');
    } catch (e) {
      console.log(`  ⚠️  target_market_idx: ${e.message}`);
    }

    // Verify
    const finalCount = await indexCollection.countDocuments();
    const indexes = await indexCollection.indexes();
    console.log(`\n✅ DONE! ${INDEX_COLLECTION} collection created with ${finalCount} documents and ${indexes.length} indexes`);
    console.log('Indexes:', indexes.map(i => i.name).join(', '));

    // Sample query to verify speed
    console.log('\n🧪 Testing search speed...');
    const start = Date.now();
    const testResults = await indexCollection.find({
      address: { $regex: /^5 FULHAM/i }
    }).limit(10).toArray();
    const elapsed = Date.now() - start;
    console.log(`  Query "5 FULHAM" returned ${testResults.length} results in ${elapsed}ms`);
    if (testResults.length > 0) {
      console.log(`  First result: ${testResults[0].address} (${testResults[0].suburb})`);
    }

  } catch (error) {
    console.error('❌ Error:', error);
    process.exit(1);
  } finally {
    await client.close();
    console.log('\n🔌 Connection closed');
  }
}

main();
